src/CRM/CivixBundle/Resources/Example/SchemaBuilderTrait.php

```diff
@@ -0,0 +1,256 @@
+        $xmlSchemas = glob($absXmlSchemaGlob);
+
+        if (!count($xmlSchemas)) {
+          throw new \Exception("Could not find files matching '$xmlSchemaGlob'. You may want to run `civix generate:entity` before running this command.");
+        }
+
+        foreach ($xmlSchemas as $xmlSchema) {
+          $dom = new \DomDocument();
+          $xmlString = file_get_contents($xmlSchema);
+          $dom->loadXML($xmlString);
+          $xml = simplexml_import_dom($dom);
+          if (!$xml) {
+            throw new \Exception('There is an error in the XML for ' . $xmlSchema);
+          }
+          $this->specification->getTable($xml, $this->database, $this->tables);
+          $name = (string) $xml->name;
+          $this->tables[$name]['name'] = $name;
+          $sourcePath = substr($xmlSchema, strlen("{$this->basedir}/"));
+          $this->tables[$name]['sourceFile'] = $this->sourceFilePrefix . $sourcePath;
+        }
+
+        $this->orderTables($this->tables);
+        $this->resolveForeignKeys($this->tables);
+
+        return $this;
+      }
+
+      /**
+       * Create DAO files.
+       *
+       * @return $this
+       */
+      public function generateDaoFiles() {
+        $config = new \stdClass();
+        $config->phpCodePath = "{$this->basedir}/";
+        //    $config->sqlCodePath = $basedir->string('sql/');
+        $config->tables = $this->tables;
+
+        foreach ($this->tables as $table) {
+          $dao = new \CRM_Core_CodeGen_DAO($config, (string) $table['name'], $this->tsFunction);
+          // Don't display gencode's output
+          ob_start();
+          $dao->run();
+          ob_end_clean();
+          $file = $this->basedir . "/{$table['base']}{$table['fileName']}";
+          $this->files[] = ['type' => 'dao', 'file' => $file];
+        }
+
+        return $this;
+      }
+
+      /**
+       * @param string $type
+       *   The type of SQL file to generate, 'CREATE' or 'DROP'.
+       *
+       * @return string
+       * @throws \Exception
+       */
+      public function generateSql($type) {
+        $generators = ['CREATE' => 'generateCreateSql', 'DROP' => 'generateDropSql'];
+        $generator = $generators[strtoupper($type)] ?? NULL;
+        if (!$generator) {
+          throw new \Exception("Invalid SQL content type: $type");
+        }
+
+        $tmpFile = sys_get_temp_dir() . DIRECTORY_SEPARATOR . bin2hex(random_bytes(12));
+        $config = new \stdClass();
+        $config->sqlCodePath = dirname($tmpFile);
+        $config->tables = $this->tables;
+
+        $schema = new \CRM_Core_CodeGen_Schema($config);
+
+        // We're poking into an internal class+function (`$schema->$generator()`) that changed in v5.23.
+        // Beginning in 5.23: $schema->$function() returns an array with file content.
+        // Before 5.23: $schema->$function($fileName) creates $fileName and returns void.
+        if (version_compare(\CRM_Utils_System::version(), '5.23.alpha1', '>=')) {
```

Comment:
Do we really need a compatibility layer for 5.23? Given that 5.27 is ESR and 5.28 is security release

Reply:
@mattwire Yeah, it seems borderline to me.

* Agree - the fact that 5.23 predates the current 5.27 ESR is an indicator that it can be removed. 
* OTOH, it 5.23 is less than a year old (March 2020).
* Removing the guard would mean that an extension using `generate:entity-boilerplate` would need to bump up its minimum. (The current default is `<compatibility><ver>5.0</ver></compatibility>`.)
* A lot of folks who write extensions are a few versions back.
* Basically, from my POV, the existing code already showed how to support <5.23, and it seemed easier to keep the compatibility than manage the removal.
* Then again, it should be removed at some point. If not now - then when?
* Updating `civix.git` doesn't automatically break existing extensions. The downstream dev needs to get the newer version and run it. If you need backward compat and if you're conscientious, then there are ways to proceed (e.g. stick to an old version of civix; e.g. play close attention to the diff's).
* It's not easy to download an older version of `civix`. There's only one published URL (`https://download.civicrm.org/civix/civix.phar`) which is evergreen. The versioned URLs aren't publicly visible.
* If we published/hyperlinked the older builds, then maybe we could be more aggressive about phase-outs within `civix`.

Reply:
I'm not sure I would have written the support for pre 5.23 - but given that it is written I think we should keep for now